File: app/src/androidTest/java/io/ffem/reports/ui/MainActivityTest.java
package io.ffem.reports.ui;


import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewParent;

import org.hamcrest.Description;
import org.hamcrest.Matcher;
import org.hamcrest.TypeSafeMatcher;
import org.hamcrest.core.IsInstanceOf;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import androidx.test.espresso.ViewInteraction;
import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.filters.LargeTest;
import androidx.test.platform.app.InstrumentationRegistry;
import androidx.test.rule.ActivityTestRule;
import io.ffem.reports.R;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.Espresso.pressBack;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withClassName;
import static androidx.test.espresso.matcher.ViewMatchers.withContentDescription;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static org.hamcrest.Matchers.allOf;
import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertEquals;

@LargeTest
@RunWith(AndroidJUnit4.class)
public class MainActivityTest {

    @Rule
    public ActivityTestRule<MainActivity> mActivityTestRule = new ActivityTestRule<>(MainActivity.class);

    private static Matcher<View> childAtPosition(
            final Matcher<View> parentMatcher, final int position) {

        return new TypeSafeMatcher<View>() {
            @Override
            public void describeTo(Description description) {
                description.appendText("Child at position " + position + " in parent ");
                parentMatcher.describeTo(description);
            }

            @Override
            public boolean matchesSafely(View view) {
                ViewParent parent = view.getParent();
                return parent instanceof ViewGroup && parentMatcher.matches(parent)
                        && view.equals(((ViewGroup) parent).getChildAt(position));
            }
        };
    }

    @Test
    public void useAppContext() {
        // Context of the app under test.
        Context appContext = InstrumentationRegistry.getInstrumentation().getTargetContext();

        assertEquals("io.ffem.reports", appContext.getPackageName());
    }

    @Test
    public void mainActivityTest() {
        ViewInteraction appCompatImageButton = onView(
                allOf(withId(R.id.actionInfo), withContentDescription("Info"),
                        childAtPosition(
                                allOf(withId(R.id.layout),
                                        childAtPosition(
                                                withId(android.R.id.content),
                                                0)),
                                3),
                        isDisplayed()));
        appCompatImageButton.perform(click());

        ViewInteraction appCompatTextView = onView(
                allOf(withId(R.id.textLinkSoftwareNotices), withText("Legal Information"),
                        childAtPosition(
                                childAtPosition(
                                        withClassName(is("android.widget.LinearLayout")),
                                        1),
                                4),
                        isDisplayed()));
        appCompatTextView.perform(click());

        ViewInteraction appCompatImageButton2 = onView(
                allOf(withContentDescription("Navigate Up"),
                        childAtPosition(
                                childAtPosition(
                                        withClassName(is("android.widget.LinearLayout")),
                                        0),
                                0),
                        isDisplayed()));
        appCompatImageButton2.perform(click());

        ViewInteraction appCompatTextView2 = onView(
                allOf(withId(R.id.textLinkSoftwareNotices), withText("Legal Information"),
                        childAtPosition(
                                childAtPosition(
                                        withClassName(is("android.widget.LinearLayout")),
                                        1),
                                4),
                        isDisplayed()));
        appCompatTextView2.perform(click());

        ViewInteraction textView = onView(
                allOf(withText("This software incorporates material from the projects listed below"),
                        childAtPosition(
                                allOf(withId(R.id.about_container),
                                        childAtPosition(
                                                IsInstanceOf.instanceOf(android.widget.ScrollView.class),
                                                0)),
                                0),
                        isDisplayed()));
        textView.check(matches(withText("This software incorporates material from the projects listed below")));

        ViewInteraction textView2 = onView(
                allOf(withText("Apache License 2.0"),
                        childAtPosition(
                                allOf(withId(R.id.about_container),
                                        childAtPosition(
                                                IsInstanceOf.instanceOf(android.widget.ScrollView.class),
                                                0)),
                                1),
                        isDisplayed()));
        textView2.check(matches(withText("Apache License 2.0")));

        ViewInteraction textView3 = onView(
                allOf(withText("CC BY 4.0 International"),
                        childAtPosition(
                                allOf(withId(R.id.about_container),
                                        childAtPosition(
                                                IsInstanceOf.instanceOf(android.widget.ScrollView.class),
                                                0)),
                                3),
                        isDisplayed()));
        textView3.check(matches(withText("CC BY 4.0 International")));

        ViewInteraction textView4 = onView(
                allOf(withText("Eclipse Public License 1.0"),
                        childAtPosition(
                                allOf(withId(R.id.about_container),
                                        childAtPosition(
                                                IsInstanceOf.instanceOf(android.widget.ScrollView.class),
                                                0)),
                                5),
                        isDisplayed()));
        textView4.check(matches(withText("Eclipse Public License 1.0")));

        ViewInteraction textView5 = onView(
                allOf(withText("MIT License"),
                        childAtPosition(
                                allOf(withId(R.id.about_container),
                                        childAtPosition(
                                                IsInstanceOf.instanceOf(android.widget.ScrollView.class),
                                                0)),
                                7),
                        isDisplayed()));
        textView5.check(matches(withText("MIT License")));

        ViewInteraction textView6 = onView(
                allOf(withText("Robolectric"),
                        childAtPosition(
                                allOf(withId(R.id.about_container),
                                        childAtPosition(
                                                IsInstanceOf.instanceOf(android.widget.ScrollView.class),
                                                0)),
                                8),
                        isDisplayed()));
        textView6.check(matches(withText("Robolectric")));

        ViewInteraction textView7 = onView(
                allOf(withText("Material Icons"),
                        childAtPosition(
                                allOf(withId(R.id.about_container),
                                        childAtPosition(
                                                IsInstanceOf.instanceOf(android.widget.ScrollView.class),
                                                0)),
                                4),
                        isDisplayed()));
        textView7.check(matches(withText("Material Icons")));

        ViewInteraction textView8 = onView(
                allOf(withText("Android Open Source Project  •  AssertJ  •  Gson  •  Timber  •  "),
                        childAtPosition(
                                allOf(withId(R.id.about_container),
                                        childAtPosition(
                                                IsInstanceOf.instanceOf(android.widget.ScrollView.class),
                                                0)),
                                2),
                        isDisplayed()));
        textView8.check(matches(withText("Android Open Source Project  •  AssertJ  •  Gson  •  Timber  •  ")));

        ViewInteraction appCompatImageButton3 = onView(
                allOf(withContentDescription("Navigate Up"),
                        childAtPosition(
                                childAtPosition(
                                        withClassName(is("android.widget.LinearLayout")),
                                        0),
                                0),
                        isDisplayed()));
        appCompatImageButton3.perform(click());

        pressBack();

        ViewInteraction button = onView(
                allOf(withId(R.id.button_ok),
                        childAtPosition(
                                allOf(withId(R.id.layout),
                                        childAtPosition(
                                                withId(android.R.id.content),
                                                0)),
                                4),
                        isDisplayed()));
        button.check(matches(isDisplayed()));

        ViewInteraction appCompatButton = onView(
                allOf(withId(R.id.button_ok), withText("Launch ffem Collect"),
                        childAtPosition(
                                allOf(withId(R.id.layout),
                                        childAtPosition(
                                                withId(android.R.id.content),
                                                0)),
                                4),
                        isDisplayed()));
        appCompatButton.perform(click());
    }
}

File: app/src/androidTest/java/io/ffem/reports/ui/MainActivityTest.kt
package io.ffem.reports.ui

import android.widget.ScrollView
import androidx.test.espresso.Espresso
import androidx.test.espresso.action.ViewActions
import androidx.test.espresso.assertion.ViewAssertions
import androidx.test.espresso.matcher.ViewMatchers
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.rule.ActivityTestRule
import io.ffem.reports.R
import io.ffem.reports.ui.MainActivity
import io.ffem.reports.util.TestUtil
import org.hamcrest.Matchers
import org.hamcrest.core.IsInstanceOf
import org.junit.Assert
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class MainActivityTest {
    @Rule
    var mActivityTestRule = ActivityTestRule(MainActivity::class.java)
    @Test
    fun useAppContext() {
        // Context of the app under test.
        val appContext = InstrumentationRegistry.getInstrumentation().targetContext
        Assert.assertEquals("io.ffem.reports", appContext.packageName)
    }

    @Test
    fun mainActivityTest() {
        val actionMenuItemView = Espresso.onView(
                Matchers.allOf(ViewMatchers.withId(R.id.actionInfo), ViewMatchers.withContentDescription("Info"),
                        TestUtil.childAtPosition(
                                TestUtil.childAtPosition(
                                        ViewMatchers.withId(R.id.toolbar),
                                        1),
                                0),
                        ViewMatchers.isDisplayed()))
        actionMenuItemView.perform(ViewActions.click())
        val appCompatTextView = Espresso.onView(
                Matchers.allOf(ViewMatchers.withId(R.id.textLinkSoftwareNotices), ViewMatchers.withText("Legal Information"),
                        TestUtil.childAtPosition(
                                TestUtil.childAtPosition(
                                        ViewMatchers.withClassName(Matchers.`is`("android.widget.LinearLayout")),
                                        1),
                                4),
                        ViewMatchers.isDisplayed()))
        appCompatTextView.perform(ViewActions.click())
        val appCompatImageButton2 = Espresso.onView(
                Matchers.allOf(ViewMatchers.withContentDescription("Navigate Up"),
                        TestUtil.childAtPosition(
                                TestUtil.childAtPosition(
                                        ViewMatchers.withClassName(Matchers.`is`("android.widget.LinearLayout")),
                                        0),
                                0),
                        ViewMatchers.isDisplayed()))
        appCompatImageButton2.perform(ViewActions.click())
        val appCompatTextView2 = Espresso.onView(
                Matchers.allOf(ViewMatchers.withId(R.id.textLinkSoftwareNotices), ViewMatchers.withText("Legal Information"),
                        TestUtil.childAtPosition(
                                TestUtil.childAtPosition(
                                        ViewMatchers.withClassName(Matchers.`is`("android.widget.LinearLayout")),
                                        1),
                                4),
                        ViewMatchers.isDisplayed()))
        appCompatTextView2.perform(ViewActions.click())
        val textView = Espresso.onView(
                Matchers.allOf(ViewMatchers.withText("This software incorporates material from the projects listed below"),
                        TestUtil.childAtPosition(
                                Matchers.allOf(ViewMatchers.withId(R.id.about_container),
                                        TestUtil.childAtPosition(
                                                IsInstanceOf.instanceOf(ScrollView::class.java),
                                                0)),
                                0),
                        ViewMatchers.isDisplayed()))
        textView.check(ViewAssertions.matches(ViewMatchers.withText("This software incorporates material from the projects listed below")))
        val textView2 = Espresso.onView(
                Matchers.allOf(ViewMatchers.withText("Apache License 2.0"),
                        TestUtil.childAtPosition(
                                Matchers.allOf(ViewMatchers.withId(R.id.about_container),
                                        TestUtil.childAtPosition(
                                                IsInstanceOf.instanceOf(ScrollView::class.java),
                                                0)),
                                1),
                        ViewMatchers.isDisplayed()))
        textView2.check(ViewAssertions.matches(ViewMatchers.withText("Apache License 2.0")))
        val textView3 = Espresso.onView(
                Matchers.allOf(ViewMatchers.withText("CC BY 4.0 International"),
                        TestUtil.childAtPosition(
                                Matchers.allOf(ViewMatchers.withId(R.id.about_container),
                                        TestUtil.childAtPosition(
                                                IsInstanceOf.instanceOf(ScrollView::class.java),
                                                0)),
                                3),
                        ViewMatchers.isDisplayed()))
        textView3.check(ViewAssertions.matches(ViewMatchers.withText("CC BY 4.0 International")))
        val textView4 = Espresso.onView(
                Matchers.allOf(ViewMatchers.withText("Eclipse Public License 1.0"),
                        TestUtil.childAtPosition(
                                Matchers.allOf(ViewMatchers.withId(R.id.about_container),
                                        TestUtil.childAtPosition(
                                                IsInstanceOf.instanceOf(ScrollView::class.java),
                                                0)),
                                5),
                        ViewMatchers.isDisplayed()))
        textView4.check(ViewAssertions.matches(ViewMatchers.withText("Eclipse Public License 1.0")))
        val textView5 = Espresso.onView(
                Matchers.allOf(ViewMatchers.withText("MIT License"),
                        TestUtil.childAtPosition(
                                Matchers.allOf(ViewMatchers.withId(R.id.about_container),
                                        TestUtil.childAtPosition(
                                                IsInstanceOf.instanceOf(ScrollView::class.java),
                                                0)),
                                7),
                        ViewMatchers.isDisplayed()))
        textView5.check(ViewAssertions.matches(ViewMatchers.withText("MIT License")))
        val textView6 = Espresso.onView(
                Matchers.allOf(ViewMatchers.withText("Robolectric"),
                        TestUtil.childAtPosition(
                                Matchers.allOf(ViewMatchers.withId(R.id.about_container),
                                        TestUtil.childAtPosition(
                                                IsInstanceOf.instanceOf(ScrollView::class.java),
                                                0)),
                                8),
                        ViewMatchers.isDisplayed()))
        textView6.check(ViewAssertions.matches(ViewMatchers.withText("Robolectric")))
        val textView7 = Espresso.onView(
                Matchers.allOf(ViewMatchers.withText("Material Icons"),
                        TestUtil.childAtPosition(
                                Matchers.allOf(ViewMatchers.withId(R.id.about_container),
                                        TestUtil.childAtPosition(
                                                IsInstanceOf.instanceOf(ScrollView::class.java),
                                                0)),
                                4),
                        ViewMatchers.isDisplayed()))
        textView7.check(ViewAssertions.matches(ViewMatchers.withText("Material Icons")))
        val textView8 = Espresso.onView(
                Matchers.allOf(ViewMatchers.withText("Android Open Source Project  •  AssertJ  •  Gson  •  Timber  •  "),
                        TestUtil.childAtPosition(
                                Matchers.allOf(ViewMatchers.withId(R.id.about_container),
                                        TestUtil.childAtPosition(
                                                IsInstanceOf.instanceOf(ScrollView::class.java),
                                                0)),
                                2),
                        ViewMatchers.isDisplayed()))
        textView8.check(ViewAssertions.matches(ViewMatchers.withText("Android Open Source Project  •  AssertJ  •  Gson  •  Timber  •  ")))
        Espresso.pressBack()
        val appCompatImageButton = Espresso.onView(
                Matchers.allOf(ViewMatchers.withContentDescription("Navigate up"),
                        TestUtil.childAtPosition(
                                Matchers.allOf(ViewMatchers.withId(R.id.toolbar),
                                        TestUtil.childAtPosition(
                                                ViewMatchers.withClassName(Matchers.`is`("android.widget.LinearLayout")),
                                                0)),
                                1),
                        ViewMatchers.isDisplayed()))
        appCompatImageButton.perform(ViewActions.click())
        val appCompatButton = Espresso.onView(
                Matchers.allOf(ViewMatchers.withId(R.id.button_ok), ViewMatchers.withText("Launch ffem Collect"),
                        TestUtil.childAtPosition(
                                TestUtil.childAtPosition(
                                        ViewMatchers.withId(R.id.layout),
                                        1),
                                3),
                        ViewMatchers.isDisplayed()))
        appCompatButton.perform(ViewActions.click())
    }
}